mesonbuild/modules/fs.py

```diff
@@ -254,6 +268,47 @@ def read(self, state: 'ModuleState', args: T.Tuple['FileOrString'], kwargs: 'Rea
         self.interpreter.add_build_def_file(path)
         return data
 
+    @FeatureNew('fs.copyfile', '0.64.0')
+    @typed_pos_args('fs.copyfile', (File, str), optargs=[str])
+    @typed_kwargs(
+        'fs.copyfile',
+        INSTALL_KW,
+        INSTALL_MODE_KW,
+        INSTALL_TAG_KW,
+        KwargInfo('install_dir', (str, NoneType)),
+    )
+    def copyfile(self, state: ModuleState, args: T.Tuple[FileOrString, T.Optional[str]],
+                 kwargs: CopyKw) -> ModuleReturnValue:
+        """Copy a file into the build directory at build time."""
+        if kwargs['install'] and not kwargs['install_dir']:
+            raise InvalidArguments('"install_dir" must be specified when "install" is true')
+
+        src = self.interpreter.source_strings_to_files([args[0]], strict=True)[0]
+
+        # The input is allowed to have path separators, but the output may not,
+        # so use the basename for the default case
+        dest = args[1] if args[1] else os.path.basename(src.fname)
+        if has_path_sep(dest):
+            raise InvalidArguments('Destination path may not have path seperators')
```

Comment:
Shouldn't this be a validator in the KwargInfo?

Reply:
Also "seperators" -> "separators".

Reply:
It's a positional argument, so no validators. I've thought about adding a `PosArgInfo` class to allow some of this to be handled int `typed_pos_args` because of this exact situation

Reply:
Actually, what I'd really like to do is replace both `typed_kwargs` and `typed_pos_args` with an `interface` decorator that also handles added/deprecated all in one decorator, but I have enough big projects going on atm...